Data/Text/ICU/Detect/Internal.hsc

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE EmptyDataDecls, ForeignFunctionInterface, RecordWildCards #-}
+
+module Data.Text.ICU.Detect.Internal
+    (
+      -- * Unicode charset detection API
+      CharsetDetector (..),
+      UCharsetDetector,
+      CharsetMatch(..),
+      UCharsetMatch,
+      getConfidence,
+      getName,
+      getLanguage,
+      detect
+     ) where
+
+import Data.Int (Int32)
+import Foreign.ForeignPtr (ForeignPtr, withForeignPtr)
+import Foreign.Ptr (Ptr)
+import Data.IORef (IORef)
+import Data.ByteString (ByteString)
+import Data.Typeable (Typeable)
+import Data.Text.ICU.Error.Internal (UErrorCode, handleError)
+import Foreign.C.String (CString, peekCString)
+import System.IO.Unsafe (unsafePerformIO)
+
+data CharsetDetector a = CD {
+  cdBytestring :: IORef ByteString,
+  cdEncoding :: IORef ByteString,
+  cdMatch :: IORef CharsetMatch,
+  cdStatus :: Int32 -> a,
+  cdDetector :: ForeignPtr UCharsetDetector
+}
+
+data UCharsetDetector
+
+data CharsetMatch = CM { cmMatch :: !(ForeignPtr UCharsetMatch) }
+  deriving (Eq, Typeable)
+
+data UCharsetMatch
+
+instance Show CharsetMatch where
+  show c = "CharsetMatch " ++ show (unsafePerformIO $ getName c)
+
+
+detect :: CharsetDetector a -> IO String
```

Comment:
Yeah, I think you’re right. FWIW, I meant less that `detect` should return a `Converter` and more that we might want to compose with `Converter`’s functionality rather than wrapping it (some sort of `detectAndConvert` function or something).